shell/packages/accounts-identity/accounts-identity-server.js

```diff
@@ -0,0 +1,198 @@
+  },
+
+  checkForLinkedAccounts: function() {
+    var user = Meteor.user();
+    if (user.loginIdentities) return {alreadyAccount: true};
+
+    var loginAccount = Meteor.users.findOne({"loginIdentities.id": user._id},
+                                            {fields: {_id: 1, "loginIdentities.$": 1}});
+
+    if (loginAccount) {
+      return {loginAccountId: loginAccount._id};
+    }
+
+    var nonloginAccounts = Meteor.users.find({"nonloginIdentities.id": user._id}).fetch();
+
+    var accountUserId;
+    if (nonloginAccounts.length == 0) {
+      // Make a new account for this user.
+      var newUser = {loginIdentities: [{id: user._id}],
+                     nonloginIdentities: []};
+      if (user.services.dev) {
+        newUser.signupKey = "devAccounts";
+        if (user.services.dev.isAdmin) {
+          newUser.isAdmin = true;
+        }
+        if (user.services.dev.hasCompletedSignup) {
+          newUser.hasCompletedSignup = true;
+        }
+      } else if (user.expires) {
+        // Demo user.
+        newUser.expires = user.expires;
+      }
+      var options = {};
+      accountUserId = Accounts.insertUserDoc(options, newUser);
+      return {loginAccountId: accountUserId};
+    } else {
+      var resultData = [];
+      nonloginAccounts.forEach(function(account) {
+        if (account.loginIdentities.length > 0) {
+          var loginIdentityUser =
+              Meteor.users.findOne({_id: account.loginIdentities[0].id});
+          if (loginIdentityUser) {
+            var userWithDefaults = SandstormDb.getUserIdentities(loginIdentityUser)[0];
+            resultData.push({accountId: account._id,
+                             loginIdentityUser: _.pick(userWithDefaults, "_id", "profile")});
+          }
+        }
+      });
+      return {nonloginAccounts: resultData};
+    }
+  },
+
+  linkIdentityToAccount: function(token) {
+    // Links the identity of the current user to the account that has `token` as a resume token.
+    // If the account is a demo account, attempts to gives the identity login access to the account.
+
+    check(token, String);
+
+    if (!this.userId) {
+      throw new Meteor.Error(403, "Cannot link to account if not logged in.");
+    }
+    var hashed = Accounts._hashLoginToken(token);
+    var accountUser = Meteor.users.findOne({"services.resume.loginTokens.hashedToken": hashed});
+
+    if (!accountUser) {
+      throw new Meteor.Error(404, "No account found for token: " + token);
+    }
+
+    if (accountUser.identity) {
+      throw new Meteor.Error(400, "Cannot link to an account that has an identity.");
+    }
+
+    var identityUser = Meteor.user();
+
+    if (!identityUser.profile) {
+      throw new Meteor.Error(400, "Current user is not an identity user.");
+    }
+
+    var modifier;
+    if (accountUser.expires) {
+      if (Meteor.users.findOne({"loginIdentities.id": identityUser._id})) {
```

Comment:
I would argue that "best effort" is sufficient for enforcing this constraint. The idea here is that if a user is trying to add an identity to their account which is already a login identity for another account, we're guessing that they're confused and this isn't really want they wanted to do. For example, perhaps they forgot that the other account exists, or perhaps they were expecting the accounts to merge.

I don't think we're trying to defend against an attack where an attacker tries to add the identity at exactly the same time that the user creates their account.